Hypothesis testing is one of the most common applications in statistical analysis.  We construct hypotheses in attempt to provide evidence in order to justify an action or change a belief.  We do this by forming an opinion we are prepared to believe in the absence of data and attempt to find evidence that would support changing that opinion.  

For example, we might consider launching a new product, but we would only do so if there were sufficient demand.  We would assume there was not sufficient demand and then look for evidence that there was.  If we found sufficient evidence, we would conclude that it was worth launching the new product; if we did not, we would conclude that there was not sufficient evidence to justify launching the new product.  

**Hypothesis tests can be simple or complicated; they can have multiple simultaneous components.  Python makes it simple to run either type!**

# Import data and run regression

The data can be found in the structural change data tab in the Cheat_Sheet_Hypothesis_Testing_V1_0 spreadsheet.

In [14]:
#Import the required libraries
import pandas as pd
import os.path as osp
import statsmodels.api as sm
from statsmodels.formula.api import ols
import os

#Build the path for the data file
db_dir = os.getcwd()
data_path = db_dir+r'/Cheat_Sheet_Hypothesis_Testing_V1_0.xlsx'
print(data_path)
#Use the read_excel function to pull data from the Hypothesis Testing Sheet
data = pd.read_excel(data_path,sheet_name=0,index_col='Obs')

#Build our model using statsmodels
model = ols('Y ~ X1 + X2 + X3 + X4 + X5',data).fit()
model.summary()

/Users/anthonyramelo/Library/CloudStorage/OneDrive-Queen'sUniversity/860Acquisition and Management of Data/Misc/Cheat_Sheet_Hypothesis_Testing_V1_0.xlsx


<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      Y   R-squared:                       0.867
Model:                            OLS   Adj. R-squared:                  0.863
Method:                 Least Squares   F-statistic:                     251.9
Date:                Wed, 10 Jul 2024   Prob (F-statistic):           8.52e-83
Time:                        17:59:23   Log-Likelihood:                -1029.1
No. Observations:                 200   AIC:                             2070.
Df Residuals:                     194   BIC:                             2090.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept    114.4941     14.438      7.930      0.000      86.019     142.969
X1             9.6945      5.990      1.619      0.107      -2.118      21.507
X2             0.0037      0.000      8.059      0.000       0.003       0.005
X3             0.0224      0.011      2.102      0.037       0.001       0.043
X4             0.1051      0.003     33.691      0.000       0.099       0.111
X5            -9.9705      6.454     -1.545      0.124     -22.699       2.758
==============================================================================
Omnibus:                        4.399   Durbin-Watson:                   1.280
Prob(Omnibus):                  0.111   Jarque-Bera (JB):                3.228
Skew:                           0.171   Prob(JB):                        0.199
Kurtosis:                       2.480   Cond. No.                     6.33e+04
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 6.33e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

# Performing Hypothesis Testing using the Wald Test

Using the following arguments, we will test and see whether $X_1$, $X_2$, and $X_3$ have a coefficient of zero in front of them. In other words, we are testing the correllation.

In [15]:
'''
We put our null hypothesis in the following string format. Here we are going
to test whether X1, X2, and X3 have a 0 coefficient in front of them.
'''
hypothesis = '(X1=0, X2=0, X3=0)'

In [16]:
#Pass the hypothesis to the Wald_Test
print(model.wald_test(hypothesis))

<F test: F=array([[25.19193307]]), p=8.273060507538756e-14, df_denom=194, df_num=3>


/opt/anaconda3/lib/python3.11/site-packages/statsmodels/base/model.py:1906: FutureWarning: The behavior of wald_test will change after 0.14 to returning scalar test statistic values. To get the future behavior now, set scalar to True. To silence this message while retaining the legacy behavior, set scalar to False.
  warnings.warn(


The p-value above is what we are concerned with the most. You should have gotten a value of 8.27306050734528e-14 (the 'e' here means $\times 10 \text{ to the power of}$). For example, 8.27e-14 is $8.27\times 10^{-14}$ (very small!). If our p-value is less than $0.05$, which it clearly is here, then we **reject the null hypothesis**. This tells us that there is, to great confidence, some correllation between the three parameters and our $Y$ value. Let's try another example where we test the correllation of just $X_1$ and $X_5$.

In [17]:
hypothesis_2 = '(X1=0,X5=0)'
print(model.wald_test(hypothesis_2))

<F test: F=array([[2.442931]]), p=0.08957531437703686, df_denom=194, df_num=2>


Here, our p-value is much higher, about $0.09$. This p-value is not less than $0.05$, so in this case, we **do not reject the null hypothesis**. Note, this does not mean that we have confirmed $X_1$ and $X_5$ are not correllated with our $Y$ value, we simply found that there is not enough evidence to suggest that they are.

# General Notes

- All of these restrictions are tested simultaneously as part of the null hypothesis.
- You can test many null hypotheses using the linearHypothesis command.
- There are other tests (for heteroskedasticity, specification error, etc.) that cannot be tested using this approach.